bundles/org.openhab.binding.mielecloud/README.md

```diff
@@ -0,0 +1,634 @@
+It may take some time until the registration e-mail arrives.
+
+There is no auto discovery for the Miele cloud account.
+The account is paired using OAuth2 with your Miele login and the developer credentials obtained from the [Miele Developer Portal](https://www.miele.com/f/com/en/register_api.aspx).
+To pair the account go to the binding's configuration UI at `https://<your openHAB address>/mielecloud`.
+For a standard openHABian Pi installation the address is [https://openhabianpi:8443/mielecloud](https://openhabianpi:8443/mielecloud).
+Note that your browser will file a warning that the certificate is self-signed.
+This is fine and you can safely continue.
+It is also possible to use an unsecured connection for pairing but it is strongly recommended to use a secured connection because your credentials will otherwise be transferred without encryption over the local network.
+For more information on this topic, see [Securing access to openHAB](https://www.openhab.org/docs/installation/security.html#encrypted-communication).
+For a detailed walk through the account configuration, see [Account Configuration Example](#account-configuration-example).
+
+Once a Miele account is paired, all supported appliances are automatically discovered as individual things and placed in the inbox.
+They can then be paired with your favorite management UI.
+As an alternative, the binding configuration UI provides a things-file template per paired account that can be used to pair the appliances.
+
+## Thing Configuration
+
+A Miele cloud account needs to be configured to get access to your appliances.
+After that appliances can be configured.
+
+### Account Configuration
+
+The Miele cloud account must be paired via the binding configuration UI before a bridge that relies on it can be configured in openHAB.
+For details on the configuration UI see [Discovery](#discovery) and [Account Configuration Example](#account-configuration-example).
+The account serves as a bridge for the things representing your appliances.
+On success the configuration assistant will directly configure the account without requiring further actions.
+As an alternative, it provides a things-file template.
+
+The account has the following parameters:
+
+| Name        | Type      | Description                                                                                                                                                | Notes                                |
+| ----------- | --------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------ |
+| locale      | mandatory | The locale to use for full text channels of things from this account. Possible values are `en`, `de`, `da`, `es`, `fr`, `it`, `nl`, `nb`. Default is `en`. | Prior to the 7th January 2021 `da`, `es`, `fr`, `it`, `nl` and `nb` will default to English. |
```

Comment:
It is possible to textually configure the account but it must still be authorized via the binding's configuration UI. I will try to explain why we did it in the way we did, maybe you have suggestions for improvements.

Access to the Miele cloud account is granted via OAuth2 which 

1. requires the user to sign in on some website and
2. then transfer the tokens required to access the service to the binding in multiple steps.

Regarding 1.: The "common" way to do this in openHAB is, as our research showed, to contribute a website to the repo for openhab.github.io/auth and redirect the user there with the instruction to copy the tokens over to the openHAB instance. This seemed cumbersome and error prone to us which is why we thought of an alternative way. We register a website in the binding which serves as a configuration UI and captures all required tokens so they can be directly processed. As the login step is required anyway we didn't see any advantage in introducing another manual step. The order of creating the textual configuration and authorizing the account doesn't matter to the binding.

Regarding 2.: The openHAB OAuth2 client is used which handles the access and refresh tokens. Thus, duplicating the tokens in the thing configuration didn't seem reasonable to us. Apart from that, the tokens change in regular intervals and we do not want the users to have to swap them. The OAuth2 client needs a unique `serviceHandle` to identify the tokens. Currently the thing UID is used as we assumed that the users wouldn't change them once configured (we also warn that this shouldn't be done [in this section of the documentation](https://github.com/BjoernLange/openhab-miele-cloud-binding-beta/blob/master/Documentation.md#account-configuration-example)).

Regarding the primary suggestion: Yes, the access token is used to identify the account, but it changes over time which is why we cannot (or can we?) use it to identify the bridge. The only solution I see is to introduce an artificial key that must not be changed. What are your thoughts on this? Are there alternative solutions? Or is it fine given these circumstances?

Reply:
How long is the refreshToken valid here? Usually, it is long-living, so it serves as the credentials to a certain account, just as username+password does. We hence use refreshTokens as configuration parameters for the account in other bindings like e.g. the [Tesla binding](https://github.com/openhab/openhab-addons/tree/d6364aceb1d952e80c0c8ed758af5d06d072ffa4/bundles/org.openhab.binding.tesla#bridge-configuration).
I assume that Miele does not support the credentials grant type, so optionally allowing username+pwd as parameters is likely no option.

What you have to keep in mind: The textual configuration should allow setting up your system without any additional manual steps. Tokens stored by the OAuthClient won't be available in this situation and must not be relied upon.

It is ok for the binding to update configuration values - so if a refreshToken becomes invalid and is replaced, the configuration can be automatically updated. This will work for both UI- and textual configured things, although clearly only for the UI-configured ones it will be persisted over a restart.

Reply:
> How long is the refreshToken valid here? Usually, it is long-living, so it serves as the credentials to a certain account, just as username+password does.

Unfortunately the refreshToken changes with each token refresh, so using it is not an option either.

> I assume that Miele does not support the credentials grant type, so optionally allowing username+pwd as parameters is likely no option.

The credentials grant type is not supported, only the authorization code grant flow.

> What you have to keep in mind: The textual configuration should allow setting up your system without any additional manual steps. Tokens stored by the OAuthClient won't be available in this situation and must not be relied upon.

We are aware of this, but it is difficult to achieve without duplicating the OAuth2 tokens in the thing configuration which has the downside of missing persistence for file based configuration. The user would need to manually adapt the files or re-authorize each time openHAB is restarted. Doesn't seem like a good option to me.

> It is ok for the binding to update configuration values - so if a refreshToken becomes invalid and is replaced, the configuration can be automatically updated. This will work for both UI- and textual configured things, although clearly only for the UI-configured ones it will be persisted over a restart.

These persistence issues are the reason why we wanted to avoid replacing configuration values.

How can we solve this?

Reply:
> Unfortunately the refreshToken changes with each token refresh

Ok, then it indeed does not make any sense.
What would you suggest as a unique identifier for an account then? The most logical choice is then probably the e-mail address. Can you get hold of this during the "account pairing" process? Can it be checked that it is the right one? This would make sure that the bridges can be cleanly differentiated and that the user can tell which is which in case there are multiple accounts configured on the same system.

Reply:
> The most logical choice is then probably the e-mail address. Can you get hold of this during the "account pairing" process?

Not directly. We could ask the user to enter it but we would stand no chance to verify it.

I see two possible alternatives:
- The client ID. This is unique for each Miele Account and we already ask for it because we need it in the authorization code flow.
- Some artificial identifier. We could generate one during authorization, e.g. a UUID, or ask the user to enter one.

Reply:
Then let's go for the client ID!

Reply:
> the client ID is not unique for each account

Did they give any example/reason? I can hardly believe it as the client id looks like a random UUID that is generated when requesting a developer account.
Besides this, one has to provide e-mail address and an application name when applying for a dev account. One of these values (e-mail/appName/ClientID) MUST be used to identify a dev account internally. I'd very much hope that we can use one of these as our id as an artificial id does not really make any sense here.

Reply:
> Did they give any example/reason?
The problem is that you can pair multiple accounts using the same client ID and client secret in one openHAB installation. If we use the client ID as "unique" identifier then users who want to pair multiple accounts also need to register multiple developer accounts.

> Besides this, one has to provide e-mail address and an application name when applying for a dev account. One of these values (e-mail/appName/ClientID) MUST be used to identify a dev account internally. I'd very much hope that we can use one of these as our id
As far as I know the e-mail address is used internally, but we have no access to it. We neither have access to the app name. Isn't that the whole point of using OAuth? That clients don't have access to the login credentials? The only data we can access and that doesn't change with a token refresh are the client ID and client secret. But these are not unique if a user wishes to use the same client ID and client secret for two different accounts in the same openHAB installation.

Reply:
>  use the same client ID and client secret for two different accounts

What makes these "different accounts"? The fact that the user logs in with different credentials? So what is the identifier there? It's the e-mail address that uniquely identifies an account then, isn't it? Even if we cannot validate it, it would nonetheless be the perfect identifier to use.

> users who want to pair multiple accounts also need to register multiple developer accounts.

So do I understand it right, that a dev account / client id is not constrained to a user account, but it can be used with ANY user account for which the credentials are entered? Ok, in this case, the client id would rather be a configuration on the binding and not the thing (if it would need it as a config param at all).

Reply:
Sorry, my last message might have been a bit confusing, but you seem to have figured it out anyway...

> So do I understand it right, that a dev account / client id is not constrained to a user account, but it can be used with ANY user account for which the credentials are entered? Ok, in this case, the client id would rather be a configuration on the binding and not the thing (if it would need it as a config param at all).

Exactly. That's why there is no config param at all at the moment.

> What makes these "different accounts"? The fact that the user logs in with different credentials?

Yes.

> It's the e-mail address that uniquely identifies an account then, isn't it? Even if we cannot validate it, it would nonetheless be the perfect identifier to use.

Yes, the e-mail address is the unique "username" that's used to sign into the Miele Cloud Account when authorizing the bridge. So, in your opinion we should restrict the identifying config param to an e-mail address that is provided by the user?

Reply:
Yes, we should use the e-mail address as a config parameter as this is what uniquely identifies the account (also for the user himself). If it can be used to prefill any login form, it would be even nicer, but if not, so be it.

Reply:
Hi @kaikreuzer! I added the a-mail address as a config parameter. Please check whether this is fine now. I will address the other open discussions shortly.